site/sfguides/src/production_ready_rag_with_trulens_snowflake_cortex/production_ready_rag_trulens_snowflake_cortex.md

```diff
@@ -0,0 +1,483 @@
+        os.environ["SNOWFLAKE_DATABASE"]].schemas[
+          os.environ["SNOWFLAKE_SCHEMA"]].cortex_search_services[
+            os.environ["SNOWFLAKE_CORTEX_SEARCH_SERVICE"]]
+    resp = cortex_search_service.search(
+        query=query,
+        columns=["doc_text"],
+        limit=self._limit_to_retrieve,
+      )
+    if resp.results:
+      return [curr["doc_text"] for curr in resp.results]
+    session.close()
+```
+
+```python
+retriever = CortexSearchRetriever(session=session, limit_to_retrieve=4)
+
+retrieved_context = retriever.retrieve(query="How do I launch a streamlit app?")
+
+len(retrieved_context)
+```
+
+## Create a RAG with built-in observability
+Duration: 8
+
+Now that we've set up the components we need from Snowflake Cortex, we can build our RAG.
+
+We'll do this by creating a custom python class with each the methods we need. We'll also add TruLens instrumentation with the `@instrument` decorator to our app.
+
+The first thing we need to do however, is to set the database connection where we'll log the traces and evaluation results from our application. This way we have a stored record that we can use to understand the app's performance. This is done when initializing `Tru`.
+
+```python
+from trulens_eval import Tru
+
+
+db_url = "snowflake://{user}:{password}@{account}/{dbname}/{schema}?warehouse={warehouse}&role={role}".format(
+  user=os.environ["SNOWFLAKE_USER"],
+  account=os.environ["SNOWFLAKE_ACCOUNT"],
+  password=os.environ["SNOWFLAKE_USER_PASSWORD"],
+  dbname=os.environ["SNOWFLAKE_DATABASE"],
+  schema=os.environ["SNOWFLAKE_SCHEMA"],
+  warehouse=os.environ["SNOWFLAKE_WAREHOUSE"],
+  role=os.environ["SNOWFLAKE_ROLE"],
+)
+
+tru = Tru(database_url=db_url)
+```
+
+Now we can construct the RAG.
+
+```python
+from trulens_eval.tru_custom_app import instrument
+
+class RAG_from_scratch:
+
+  def __init__(self):
+    self.retriever = CortexSearchRetriever(session=session, limit_to_retrieve=4)
+  @instrument
+  def retrieve_context(self, query: str) -> list:
+    """
+    Retrieve relevant text from vector store.
+    """
+    results = self.retriever.retrieve(query)
+    return results
+
+  @instrument
+  def generate_completion(self, query: str, context_str: list) -> str:
+    """
+    Generate answer from context.
+    """
+    completion = Complete("mistral-large",query)
+    return completion
+
+  @instrument
+  def query(self, query: str) -> str:
+    context_str = self.retrieve_context(query)
+    completion = self.generate_completion(query, context_str)
+    return completion
+
+rag = RAG_from_scratch()
+```
+
+After constructing the RAG, we can set up the feedback functions we want to use to evaluate the RAG.
+
+Here, we'll use the [RAG Triad](https://www.trulens.org/trulens_eval/getting_started/core_concepts/rag_triad/). The RAG triad is made up of 3 evaluations along each edge of the RAG architecture: context relevance, groundedness and answer relevance.
+
+![RAG Triad](./assets/RAG_Triad.jpg)
```

Comment:
I don't know about for you, but for me this isn't loading.

Reply:
This is true locally for other all other quickstarts too, I'll follow up with the owner to see if this is expected.